File: apps/backend/app/application/convert_font.py
"""ConvertFontUseCase — POST /convert 및 GET /jobs/{id} 의 비즈니스 로직 오케스트레이션."""

from __future__ import annotations

import asyncio
from pathlib import Path

import structlog

from app.domain import metrics_calculator
from app.domain.models import ConvertResult, ErrorInfo, Job
from app.ports.ai_engine import AIEnginePort
from app.ports.font_processor import FontProcessorPort
from app.ports.job_store import JobStorePort
from app.ports.storage import StoragePort

logger = structlog.get_logger()


class ConvertFontUseCase:
    def __init__(
        self,
        storage: StoragePort,
        ai_engine: AIEnginePort,
        font_processor: FontProcessorPort,
        job_store: JobStorePort,
        input_bucket: str,
        output_bucket: str,
        signed_url_ttl_seconds: int,
    ) -> None:
        self._storage = storage
        self._ai_engine = ai_engine
        self._font_processor = font_processor
        self._job_store = job_store
        self._input_bucket = input_bucket
        self._output_bucket = output_bucket
        self._signed_url_ttl = signed_url_ttl_seconds

    async def start_job(self, file_bytes: bytes, filename: str) -> str:
        job = await self._job_store.create()
        logger.info("job_started", job_id=job.id, filename=filename)
        asyncio.create_task(self._run_conversion(job.id, file_bytes, filename))
        return job.id

    async def get_job(self, job_id: str) -> Job | None:
        return await self._job_store.get(job_id)

    async def _run_conversion(self, job_id: str, file_bytes: bytes, filename: str) -> None:
        structlog.contextvars.bind_contextvars(job_id=job_id)
        try:
            await self._job_store.update(job_id, status="processing", stage="uploading")
            logger.info("stage_started", stage="uploading")
            key = f"{job_id}.ttf"
            await self._storage.upload(self._input_bucket, key, file_bytes)

            await self._job_store.update(job_id, stage="parsing", progress=0.2)
            logger.info("stage_started", stage="parsing")
            try:
                original_glyphs = self._font_processor.extract_glyphs(file_bytes)
            except ValueError as e:
                await self._fail(job_id, "INVALID_TTF", "parsing", "올바른 TTF 파일이 아닙니다.")
                logger.warning("parsing_failed", error=str(e))
                return

            await self._job_store.update(job_id, stage="optimizing", progress=0.4)
            logger.info("stage_started", stage="optimizing")
            try:
                optimized = await self._ai_engine.optimize(original_glyphs)
            except Exception as e:
                await self._fail(
                    job_id, "CONVERSION_FAILED", "optimizing", "변환 중 오류가 발생했습니다."
                )
                logger.exception("optimization_failed", error=str(e))
                return

            await self._job_store.update(job_id, stage="finalizing", progress=0.8)
            logger.info("stage_started", stage="finalizing")
            converted_bytes = self._font_processor.rebuild_ttf(file_bytes, optimized)
            await self._storage.upload(self._output_bucket, key, converted_bytes)

            metrics = metrics_calculator.calculate(original_glyphs, optimized)
            signed = await self._storage.signed_url(
                self._output_bucket, key, self._signed_url_ttl
            )

            stem = Path(filename).stem
            result = ConvertResult(
                ink_saving_rate=metrics.ink_saving_rate,
                carbon_reduction_g=metrics.carbon_reduction_g,
                download_url=signed.url,
                expires_at=signed.expires_at,
                original_filename=filename,
                converted_filename=f"{stem}_eco.ttf",
            )
            await self._job_store.update(
                job_id, status="done", stage=None, progress=1.0, result=result
            )
            logger.info(
                "job_done",
                ink_saving_rate=metrics.ink_saving_rate,
                carbon_reduction_g=metrics.carbon_reduction_g,
            )
        except Exception as e:
            await self._fail(
                job_id, "SERVICE_UNAVAILABLE", None, "일시적인 오류입니다. 다시 시도해주세요."
            )
            logger.exception("job_failed_unexpected", error=str(e))
        finally:
            structlog.contextvars.unbind_contextvars("job_id")

    async def _fail(
        self, job_id: str, code: str, stage: str | None, message: str
    ) -> None:
        await self._job_store.update(
            job_id,
            status="failed",
            stage=None,
            error=ErrorInfo(code=code, stage=stage, message=message),
        )

File: apps/frontend/src/mocks/convertFont.ts
// TODO(Unit 1b): 실 API 연결 시 startConversion → POST /convert fetch로 교체
//               pollJob → GET /jobs/{id} fetch로 교체
//
// 현재 mock은 백엔드가 TTF 바이너리를 직접 응답하는 구조를 임시 시뮬레이션.
// 백엔드 최종 응답 방식(직접 전송 vs GCS Signed URL)이 확정되면
// MockVariant 타입과 pollJob 반환값을 실 응답 스키마에 맞게 교체할 것.

export type MockVariant = {
  // TODO(Unit 1b): 백엔드 응답 구조 확정 후 이 필드 교체
  //   - TTF 직접 전송 방식 → ttf_blob 유지 (또는 Response.blob() 처리로)
  //   - GCS Signed URL 방식 → download_url: string 으로 교체 후 hook에서 fetch
  ttf_blob: Blob;
  ink_saving_rate: number;
  carbon_reduction_g: number;
  converted_filename: string;
};

export type MockPollResult =
  | { status: "processing"; stage: string; progress: number }
  | { status: "done"; result: { variants: MockVariant[] } }
  | { status: "failed"; message: string };

export async function startConversion(_file: File): Promise<string> {
  await delay(300);
  return "mock-job-id";
}

export async function pollJob(
  _jobId: string,
  file: File,
  attempt: number
): Promise<MockPollResult> {
  await delay(1000);
  if (attempt === 1)
    return { status: "processing", stage: "parsing", progress: 0.3 };
  if (attempt === 2)
    return { status: "processing", stage: "optimizing", progress: 0.7 };

  const variantSpecs = [
    { rate: 0.15, carbon: 7.8, suffix: "_eco_v1" },
    { rate: 0.23, carbon: 12.1, suffix: "_eco_v2" },
    { rate: 0.31, carbon: 16.4, suffix: "_eco_v3" },
    { rate: 0.38, carbon: 20.0, suffix: "_eco_v4" },
    { rate: 0.45, carbon: 23.7, suffix: "_eco_v5" },
  ];

  // TODO(Unit 1b): 실 API 연결 시 아래 blob 생성 로직 제거.
  // 현재는 원본 파일 바이트를 그대로 복사해 TTF 직접 수신을 흉내냄.
  // 실 API는 변환된 TTF를 바이너리로 응답(또는 Signed URL로 제공)하게 됨.
  const arrayBuffer = await file.arrayBuffer();
  const variants = variantSpecs.map((v) => ({
    ttf_blob: new Blob([arrayBuffer], { type: "font/ttf" }),
    ink_saving_rate: v.rate,
    carbon_reduction_g: v.carbon,
    converted_filename: file.name.replace(/\.ttf$/i, `${v.suffix}.ttf`),
  }));

  return { status: "done", result: { variants } };
}

function delay(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}
